File: Simulator-VB/Components/KeyPad.vb
Public Class KeyPad


    Sub New()

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().
        For Each item As NumericButton In Me.Controls.OfType(Of NumericButton)
            If item.Enabled Then
                AddHandler item.Click, AddressOf _Pressed
            End If
        Next

    End Sub

    Public Event Pressed As EventHandler

    Private Sub _Pressed(sender As Object, e As EventArgs)
        If String.IsNullOrEmpty(Value) OrElse Value.Length < MaxLenght Then
            Value &= DirectCast(sender, NumericButton).Text
            RaiseEvent Pressed(Me, EventArgs.Empty)
        End If
    End Sub


    Public Property MinLenght As Integer = 4
    Public Property MaxLenght As Integer = 4

    Public Property Value As String = String.Empty

    Private Sub _Disposed(sender As Object, e As EventArgs) Handles Me.Disposed
        For Each item As NumericButton In Me.Controls.OfType(Of NumericButton)
            If item.Enabled Then
                RemoveHandler item.Click, AddressOf _Pressed
            End If
        Next
    End Sub

#Region "Events"

    Public Event Cancel As EventHandler
    Public Event Fix As EventHandler
    Public Event Confirm As EventHandler
    Public Event [Error] As EventHandler

#End Region

    Private Sub ButtonCancel_Click(sender As Object, e As EventArgs) Handles ButtonCancel.Click
        RaiseEvent Cancel(Me, EventArgs.Empty)
    End Sub

    Private Sub ButtonFix_Click(sender As Object, e As EventArgs) Handles ButtonFix.Click
        If Not String.IsNullOrEmpty(Value) AndAlso Value.Length > 0 Then
            Value = Value.Substring(0, Value.Length - 1)
        End If
        RaiseEvent Fix(Me, EventArgs.Empty)
    End Sub

    Private Sub ButtonConfirm_Click(sender As Object, e As EventArgs) Handles ButtonConfirm.Click
        If Not String.IsNullOrEmpty(Value) AndAlso Value.Length <= MaxLenght AndAlso Value.Length = MinLenght Then
            RaiseEvent Confirm(Me, EventArgs.Empty)
        Else
            RaiseEvent Error(Me, EventArgs.Empty)
        End If
    End Sub

End Class

File: Simulator-CSharp/Components/KeyPad.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using ATM.Controls;

namespace ATM.Components
{
    public partial class KeyPad : UserControl
    {
        public KeyPad()
        {
            InitializeComponent();
        }

        public event EventHandler Pressed;

        private void _Pressed(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(Valor) || Valor.Length < MaxLenght)
            {
                Valor += ((NumericButton)sender).Text;
                if (Pressed != null)
                {
                    Pressed(this, EventArgs.Empty);
                }
            }
        }


        public int MinLenght { get; set; }
        public int MaxLenght { get; set; }

        public string Valor { get; set; }

        #region Events

        public event EventHandler Cancel;
        public event EventHandler Fix;
        public event EventHandler Confirm;
        public event EventHandler Error;

        #endregion

        private void OnButtonCancelClick(object sender, EventArgs e)
        {
            if (Cancel != null)
            {
                Cancel(this, EventArgs.Empty);
            }
        }

        private void OnButtonFixClick(object sender, EventArgs e)
        {
            if (!string.IsNullOrEmpty(Valor) && Valor.Length > 0)
            {
                Valor = Valor.Substring(0, Valor.Length - 1);
            }
            if (Fix != null)
            {
                Fix(this, EventArgs.Empty);
            }
        }

        private void OnButtonConfirmClick(object sender, EventArgs e)
        {
            if (!string.IsNullOrEmpty(Valor) && Valor.Length <= MaxLenght && Valor.Length == MinLenght)
            {
                Confirm(this, EventArgs.Empty);
            }
            else
            {
                Error(this, EventArgs.Empty);
            }
        }

    }
}
